src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldNotHaveIncorrectSuffix.cs

```diff
@@ -56,13 +62,219 @@ public abstract class IdentifiersShouldNotHaveIncorrectSuffixAnalyzer : Diagnost
                                                                              DiagnosticSeverity.Warning,
                                                                              isEnabledByDefault: false,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: HelpLinkUri,
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(TypeNoAlternateRule, MemberNewerVersionRule, TypeNewerVersionRule, MemberWithAlternateRule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            TypeNoAlternateRule,
+            MemberNewerVersionRule,
+            TypeNewerVersionRule,
+            MemberWithAlternateRule);
+
+        internal const string AttributeSuffix = "Attribute";
+        internal const string CollectionSuffix = "Collection";
+        internal const string DictionarySuffix = "Dictionary";
+        internal const string EventArgsSuffix = "EventArgs";
+        internal const string EventHandlerSuffix = "EventHandler";
+        internal const string ExSuffix = "Ex";
+        internal const string ExceptionSuffix = "Exception";
+        internal const string NewSuffix = "New";
+        internal const string PermissionSuffix = "Permission";
+        internal const string StreamSuffix = "Stream";
+        internal const string DelegateSuffix = "Delegate";
+        internal const string EnumSuffix = "Enum";
+        internal const string ImplSuffix = "Impl";
+        internal const string CoreSuffix = "Core";
+        internal const string QueueSuffix = "Queue";
+        internal const string StackSuffix = "Stack";
+
+        // Dictionary that maps from a type name suffix to the set of base types from which
+        // a type with that suffix is permitted to derive.
+        private static readonly ImmutableDictionary<string, ImmutableArray<string>> s_suffixToBaseTypeNamesDictionary = ImmutableDictionary.CreateRange(
```

Comment:
They're not quite the same:

* `DataSet`/`DataTable`: In CA1710, FxCop correctly requires the name of a `DataSet`-derived class to end in `DataSet`, but in CA1711, FxCop incorrectly does not complain if a non-`DataSet`-derived class does end in `DataSet`. So these appear in his dictionary but not in mind.

* Balaji's dictionary has the additional information of whether it's ok for a the name of a variable of each type to end in "Collection" instead of the more specific type. Mine doesn't have/need that information.

This could be DRY'd out with not too much work, but I'd rather not do it now.